File: CSharp/Core/MsLogger.Core/DataStoreLoggerProvider.cs
using System.Collections.Concurrent;
using LogViewer.Core;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace MsLogger.Core;

public class DataStoreLoggerProvider: ILoggerProvider
{

    #region Constructor
    
    public DataStoreLoggerProvider(IOptionsMonitor<DataStoreLoggerConfiguration> config, ILogDataStore dataStore)
    {
        _dataStore = dataStore;
        _currentConfig = config.CurrentValue;
        _onChangeToken = config.OnChange(updatedConfig => _currentConfig = updatedConfig);
    }

    #endregion

    #region fields
    
    private DataStoreLoggerConfiguration _currentConfig;

    private readonly IDisposable? _onChangeToken;
    protected readonly ILogDataStore _dataStore;

    protected readonly ConcurrentDictionary<string, DataStoreLogger> _loggers = new();
    
    #endregion

    #region Methods
    
    public ILogger CreateLogger(string categoryName)
        => _loggers.GetOrAdd(categoryName, name => new DataStoreLogger(name, GetCurrentConfig, _dataStore));

    protected DataStoreLoggerConfiguration GetCurrentConfig()
        => _currentConfig;

    public void Dispose()
    {
        _loggers.Clear();
        _onChangeToken?.Dispose();
    } 

    #endregion
}

File: VB/Core/MsLoggerVB.Core/DataStoreLoggerProvider.vb
Imports System.Collections.Concurrent
Imports LogViewerVB.Core
Imports Microsoft.Extensions.Logging
Imports Microsoft.Extensions.Options

Public Class DataStoreLoggerProvider : Implements ILoggerProvider

#Region "Constructors"

    Public Sub New(config As IOptionsMonitor(Of DataStoreLoggerConfiguration), dataStore As ILogDataStore)

        _dataStore = dataStore
        _currentConfig = config.CurrentValue
        _onChangeToken = config.OnChange(Sub(updatedConfig) _currentConfig = updatedConfig)

    End Sub

#End Region

#Region "Fields"

    Private _currentConfig As DataStoreLoggerConfiguration

    Private ReadOnly _onChangeToken As IDisposable
    Protected ReadOnly _dataStore As ILogDataStore

    Protected ReadOnly _loggers As ConcurrentDictionary(Of String, DataStoreLogger) =
                         New ConcurrentDictionary(Of String, DataStoreLogger)()

#End Region

#Region "Methods"

    Public Overridable Function CreateLogger(categoryName As String) As ILogger Implements ILoggerProvider.CreateLogger

        Return _loggers.GetOrAdd(categoryName, Function(name) New DataStoreLogger(name, AddressOf GetCurrentConfig, _dataStore))

    End Function

    Protected Function GetCurrentConfig() As DataStoreLoggerConfiguration

        Return _currentConfig

    End Function

    Public Sub Dispose() Implements IDisposable.Dispose

        _loggers.Clear()
        _onChangeToken?.Dispose()

    End Sub

#End Region

End Class